.github/workflows/php-qa.yml

```diff
@@ -1,77 +1,95 @@
+                        dependency-versions: 'highest'
+                        composer-options: '--ignore-platform-reqs'
+                        experimental: true
+        continue-on-error: ${{ matrix.experimental == true }}
         steps:
-            - uses: actions/checkout@v2
+            -   name: Update "USE_COVERAGE" env var based on matrix
+                if: ${{ matrix.php-ver == '7.4' && matrix.dependency-versions == 'highest' }}
+                run: echo "USE_COVERAGE=yes" >> $GITHUB_ENV
 
-            - name: Setup PHP
-              uses: shivammathur/setup-php@v2
-              with:
-                php-version: ${{ matrix.php-versions }}
-                ini-values: zend.assertions=1, error_reporting=-1, display_errors=On
-                coverage: ${{ ( matrix.php-versions == '7.4' && 'xdebug' ) || 'none' }}
-                tools: parallel-lint
+            -   name: Checkout
+                uses: actions/checkout@v2
 
-            - name: Check syntax error in sources
-              if: ${{ matrix.dependency-versions == 'highest' }}
-              run: parallel-lint ./src/ ./tests/
+            -   name: Setup PHP
+                uses: shivammathur/setup-php@v2
+                with:
+                    php-version: ${{ matrix.php-ver }}
+                    ini-values: zend.assertions=1, error_reporting=-1, display_errors=On
+                    coverage: ${{ ((env.USE_COVERAGE == 'yes') && 'xdebug') || 'none' }}
+                    tools: cs2pr
 
-            - name: Install dependencies - normal
-              if: ${{ matrix.php-versions != '8.2' }}
-              uses: "ramsey/composer-install@v2"
-              with:
-                dependency-versions: ${{ matrix.dependency-versions }}
+            -   name: Install parallel-lint
+                if: ${{ matrix.dependency-versions == 'highest' }}
+                run: composer require php-parallel-lint/php-parallel-lint:^1.3.1 --dev --no-update
 
-            - name: Install dependencies - ignore-platform-reqs
-              if: ${{ matrix.php-versions == '8.2' }}
-              uses: "ramsey/composer-install@v2"
-              with:
-                dependency-versions: ${{ matrix.dependency-versions }}
-                composer-options: "--ignore-platform-reqs"
+            -   name: Install dependencies
+                uses: ramsey/composer-install@v2
+                with:
+                    dependency-versions: ${{ matrix.dependency-versions }}
+                    composer-options: ${{ matrix.composer-options }}
 
-            - name: Check cross-version PHP compatibility
-              if: ${{ matrix.php-versions == '7.4' && matrix.dependency-versions == 'highest' }} # results is same across versions, do it once
-              run: composer phpcompat
+            -   name: Lint PHP sources
+                if: ${{ matrix.dependency-versions == 'highest' }}
+                run: ./vendor/bin/parallel-lint ./src/ --checkstyle | cs2pr
```

Comment:
Why are you no longer linting the `/tests/` directory ? If there is a parse error, the test run will error out, but why not error out early by including the test files in the linting check ?

Reply:
If there's a parse error, the failure will happen as soon as test case is loaded, so _very early_. That means that linting the tests will fail _few milliseconds_ earlier, but in the great majority of cases, when there're no syntax errors, it takes much more time to lint tests. 1 second * 8 versions * _each time_ the workflow run is *much* more time that I can save failing a little bit earlier _only if a failure is there_ (very rarely).

At work, where we pay for GitHub action time, by removing lint for the test folders we saved a non-negligible amount of money (we have _hundreds_ of repositories actively developed).